### PR TITLE
2D Conv transpose support 

### PR DESCRIPTION
**Pull request by tejank10**
This PR adds support for 2D Conv Transpose for Flux, along with [#54](https://github.com/FluxML/NNlib.jl/pull/54) in NNlib.jl

### CODE REVIEW

**Comment by tejank10** [comment]
`gradtests` are failing for `∇conv_data`. Gradtest wrt data is passing. I found that the culprit is the gradient wrt the weight which is differing drastically.

**Comment by tejank10** [comment]
NNlib counterpart of this PR needs to be merged for the checks to pass

**Comment by Aniket1998** [comment]
What's the status of this PR? Working on an InfoGAN model for the Flux model zoo, unable to write a version that would work on SVHN without conv transpose

**Comment by yuehhua** [comment]
It's there anything I can help? I'm waiting for this.

File: test/tracker.jl [hunk 1]
```diff
@@ -1,7 +1,11 @@
 using Flux
 using Flux.Tracker, Test, NNlib
+<<<<<<< HEAD
```
Comment (vchuravy) [inline test/tracker.jl]: Merge artificat

File: src/layers/conv.jl [hunk 2]
```diff
@@ -51,6 +51,53 @@ function Base.show(io::IO, l::Conv)
...
+  stride::NTuple{N,Int}
+  pad::NTuple{N,Int}
+  dilation::NTuple{N,Int}
+end
+
+ConvTranspose(w::AbstractArray{T,N}, b::AbstractVector{T}, σ = identity;
+              stride = 1, pad = 0, dilation = 1) where {T,N} =
+  ConvTranspose(σ, w, b, expand.(sub2(Val(N)), (stride, pad, dilation))...)
+
+ConvTranspose(k::NTuple{N,Integer}, ch::Pair{<:Integer,<:Integer}, σ = identity; init = initn,
```
Comment (vchuravy) [inline src/layers/conv.jl]:
```
julia> ConvTranspose((2, 2), 3=>3)
ERROR: UndefVarError: initn not defined
Stacktrace:
 [1] ConvTranspose(::Tuple{Int64,Int64}, ::Pair{Int64,Int64}, ::Function) at /home/vchuravy/.julia/packages/Flux/hguaX/src/layers/conv.jl:84 (repeats 2 times)
```

Comment (vchuravy) [reply src/layers/conv.jl]: Also:

```
julia> ConvTranspose((2, 2), 3=>64; init=Flux.glorot_uniform)(rand(4, 4, 3, 10))
ERROR: MethodError: no method matching ∇conv_data(::Array{Float64,4}, ::Array{Float32,4}; stride=(1, 1), pad=(0, 0), dilation=(1, 1))
Closest candidates are:
  ∇conv_data(::AbstractArray, ::TrackedArray; kw...) at /home/vchuravy/.julia/packages/Flux/hguaX/src/tracker/lib/array.jl:390
  ∇conv_data(::TrackedArray, ::AbstractArray; kw...) at /home/vchuravy/.julia/packages/Flux/hguaX/src/tracker/lib/array.jl:391
  ∇conv_data(::A<:AbstractArray, ::A<:AbstractArray; size, pad, stride, dilation, flipkernel) where A<:AbstractArray at /home/vchuravy/.julia/packages/NNlib/nf8OC/src/conv.jl:74
  ...
```

Comment (tejank10) [reply src/layers/conv.jl]: Thanks for pointing these out, I've fixed them in the latest commit.

**Comment by tejank10** [comment]
I just made a PR in CuArrays ([#223](https://github.com/JuliaGPU/CuArrays.jl/pull/223)) to reflect the changes in `conv` for CuArrays. With that, and latest `conv_transpose` branch of NNlib this should be working.

**Comment by NTimmons** [comment]
This looks like it never got merged. Is there no ConvTranspose in Flux?

**Comment by tejank10** [comment]
@MikeInnes if [#54](https://github.com/FluxML/NNlib.jl/pull/54) is fine then let's get it, and this PR merged?

File: src/layers/conv.jl [hunk 3]
```diff
@@ -77,6 +125,7 @@ struct DepthwiseConv{N,F,A,V}
   bias::V
   stride::NTuple{N,Int}
   pad::NTuple{N,Int}
+  dilation::NTuple{N,Int}
```
Comment (staticfloat) [inline src/layers/conv.jl]: This seems like perhaps an unrelated change?

**Comment by staticfloat** [comment]
In my limited autoencoder testing, this seems to be working, but I'm worried about the failing gradient tests.  @tejank10 do you know why they're failing?

**Comment by tejank10** [comment]
Trying tests with `convAPI` branch of `CuArrays` in [https://github.com/JuliaGPU/CuArrays.jl/pull/223](this PR) should get them passing.

**Comment by MikeInnes** [comment]
I just merged that PR. Worth testing this again, figuring out if we need a CuArrays tag etc.

**Comment by tejank10** [comment]
The `gradtests` are failing because it is using NNlib v0.4.3. Trying the tests with the `NNlib-master` seems to be working.

**Comment by MikeInnes** [comment]
You should be able to add NNlib master to the manifest and get it tested that way.

**Comment by MikeInnes** [comment]
Awesome stuff @tejank10, thanks!

**Comment by sidhantnagpal** [comment]
Thanks for working on this. It will be useful in implementing GANs using Flux.

**Comment by MikeInnes** [comment]
Also thanks a lot to @staticfloat for the review!

**Comment by AzamatB** [comment]
Does this support only 2D images, as the name suggests, or 1D, 2D as well?

**Comment by tejank10** [comment]
It should support 1D and 3D as well.
